p2p/starknet/block_bodies.go

```diff
@@ -0,0 +1,164 @@
+package starknet
+
+import (
+	"slices"
+
+	"github.com/NethermindEth/juno/adapters/core2p2p"
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/p2p/starknet/spec"
+	"github.com/NethermindEth/juno/utils"
+	"google.golang.org/protobuf/proto"
+)
+
+type streamState int
+
+const (
+	_ streamState = iota
+
+	sendDiff // initial
+	sendClasses
+	sendBlockFin
+	terminal // final state
+)
+
+type blockBodyIterator struct {
+	bcReader blockchain.Reader
+	log      utils.Logger
+
+	state       streamState
+	header      *core.Header
+	stateUpdate *core.StateUpdate
+}
+
+func newBlockBodyIterator(bcReader blockchain.Reader, header *core.Header, log utils.Logger) *blockBodyIterator {
+	return &blockBodyIterator{
+		state:       sendDiff,
+		header:      header,
+		bcReader:    bcReader,
+		log:         log,
+		stateUpdate: nil, // to be filled during one of next states
+	}
+}
+
+func (b *blockBodyIterator) hasNext() bool {
+	statesHaveTransitions := []streamState{
+		sendDiff,
+		sendClasses,
+		sendBlockFin,
+	}
+	return slices.Contains(statesHaveTransitions, b.state)
+}
+
+// Either BlockBodiesResponse_Diff, *_Classes, *_Proof, *_Fin
+func (b *blockBodyIterator) next() (msg proto.Message, valid bool) {
+	switch b.state {
+	case sendDiff:
+		msg, valid = b.diff()
+		b.state = sendClasses
+	case sendClasses:
+		msg, valid = b.classes()
+		b.state = sendBlockFin
+	case sendBlockFin:
+		msg, valid = b.fin()
+		b.state = terminal
+	case terminal:
+		panic("next called on terminal state")
+	default:
+		b.log.Errorw("Unknown state in blockBodyIterator", "state", b.state)
+	}
+
+	return
+}
+
+func (b *blockBodyIterator) classes() (proto.Message, bool) {
+	var classes []*spec.Class
+
+	state, closer, err := b.bcReader.StateAtBlockNumber(b.header.Number)
+	if err != nil {
+		return b.finAndSetTerminalState()
+	}
+	defer b.callAndLogErr(closer, "Error closing state reader in blockBodyIterator.classes()")
+
+	stateDiff := b.stateUpdate.StateDiff
+
+	for _, hash := range stateDiff.DeclaredV0Classes {
+		cls, err := state.Class(hash)
+		if err != nil {
+			return b.finAndSetTerminalState()
+		}
+
+		classes = append(classes, core2p2p.AdaptClass(cls.Class, hash))
+	}
+	for _, class := range stateDiff.DeclaredV1Classes {
+		cls, err := state.Class(class.ClassHash)
+		if err != nil {
+			return b.finAndSetTerminalState()
+		}
+
+		cairo1Cls := cls.Class.(*core.Cairo1Class)
+		classes = append(classes, core2p2p.AdaptClass(cls.Class, cairo1Cls.Hash()))
+	}
```

Comment:
I think we also need to check the `deployedContracts` because there have been cases in the older blocks where Contracts were deployed without them being declared (I don't know why, its one of those Starknet things). Please have a look at the `fetchUnknownClasses()` in `sync/sync.go`.

Reply:
I have updated the implementation to make sure there are no duplicate classes. Also, added a test case for it.